src/connection.rs

```diff
@@ -286,3 +291,142 @@ where
         }
     }
 }
+
+pub struct Select<'a> {
+    cursor: Option<Cursor<'a>>,
+    error: Option<Error>,
+    columns_map: HashMap<String, usize>,
+}
+
+impl<'a> Select<'a> {
+    pub fn query(conn: &'a Connection, query: impl AsRef<str>) -> Self {
+        match conn.prepare(query) {
+            Ok(statement) => Self {
+                error: None,
+                columns_map: Self::create_columns_map(&statement),
+                cursor: Some(statement.into_cursor()),
+            },
+            Err(err) => Self {
+                cursor: None,
+                error: Some(err),
+                columns_map: Default::default(),
+            },
+        }
+    }
+
+    fn create_columns_map(statement: &Statement) -> HashMap<String, usize> {
+        (0..statement.column_count())
+            .map(|i| (statement.column_name(i).to_string(), i))
+            .collect()
+    }
+}
+
+impl<'a> From<Statement<'a>> for Select<'a> {
+    fn from(statement: Statement<'a>) -> Self {
+        Self {
+            error: None,
+            columns_map: Self::create_columns_map(&statement),
+            cursor: Some(statement.into_cursor()),
+        }
+    }
+}
+
+impl<'a> Iterator for Select<'a> {
+    type Item = Result<Row>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let columns_map = self.columns_map.clone();
+        match (self.cursor.as_mut(), self.error.take()) {
+            (_, Some(err)) => Some(Err(err)),
+            (Some(cursor), _) => match cursor.next() {
+                Err(err) => Some(Err(err)),
+                Ok(maybe_row) => Ok(maybe_row.map(|row| Row {
+                    row: row.to_vec(),
+                    columns_map,
+                }))
+                .transpose(),
+            },
+            (None, None) => None,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Row {
+    row: Vec<Value>,
+    columns_map: HashMap<String, usize>,
+}
+
+pub trait ValueInto: Sized {
+    fn try_convert_value_into(value: &Value) -> Option<Self>;
+}
+
+pub trait ColumnIndex: std::fmt::Debug {
```

Comment:
Again, if possible, it would be good to avoid new traits.

Reply:
If not, just call it `Index`. Should probably live next to `Cursor`.

Reply:
> Again, if possible, it would be good to avoid new traits.

I don't see any other way. `std::ops::Index` would be a good candidate but it is implemented on the collection not on the index type.

> If not, just call it `Index`. Should probably live next to `Cursor`.

That would conflict with std's `Index` https://doc.rust-lang.org/std/ops/trait.Index.html